src/validators/function.rs

```diff
@@ -300,20 +323,51 @@ pub fn convert_err<'a>(py: Python<'a>, err: PyErr, input: &'a impl Input<'a>) ->
 
 #[pyclass(module = "pydantic_core._pydantic_core")]
 pub struct ValidationInfo {
-    #[pyo3(get)]
-    data: Option<Py<PyDict>>,
     #[pyo3(get)]
     config: PyObject,
     #[pyo3(get)]
     context: Option<PyObject>,
+    data: Option<Py<PyDict>>,
+    field_name: Option<String>,
 }
 
 impl ValidationInfo {
-    fn new(py: Python, extra: &Extra, config: &PyObject) -> Self {
-        Self {
-            data: extra.data.map(|v| v.into()),
-            config: config.clone_ref(py),
-            context: extra.context.map(|v| v.into()),
-        }
+    fn new(py: Python, extra: &Extra, config: &PyObject, is_model_instance_method: bool) -> PyResult<Self> {
+        let res = if is_model_instance_method {
+            let field_name = match extra.field_name {
+                Some(v) => v,
+                None => return Err(PyRuntimeError::new_err("This validator expected to be run inside the context of a model field but not model field was found")),
+            };
+            Self {
+                config: config.clone_ref(py),
+                context: extra.context.map(|v| v.into()),
+                field_name: Some(field_name.to_string()),
+                data: extra.data.map(|v| v.into()),
+            }
+        } else {
+            Self {
+                config: config.clone_ref(py),
+                context: extra.context.map(|v| v.into()),
+                field_name: None,
+                data: None,
+            }
+        };
+        Ok(res)
+    }
+}
+
+#[pymethods]
+impl ValidationInfo {
+    #[getter]
+    fn get_data(&self) -> PyResult<Py<PyDict>> {
+        self.data
+            .clone()
+            .ok_or_else(|| PyAttributeError::new_err("No attribute named 'data'"))
+    }
+    #[getter]
+    fn get_field_name(&self) -> PyResult<String> {
+        self.field_name
+            .clone()
+            .ok_or_else(|| PyAttributeError::new_err("No attribute named 'field_name'"))
```

Comment:
I think it should be more efficient to do something like
```suggestion
    fn get_field_name(&self, py: Python) -> PyResult<&PyString> {
        match self.field_name {
          Some(ref s) => Ok(PyString::new(py, s)),
          None => Err(PyAttributeError::new_err("No attribute named 'field_name'")
        }
```

Reply:
6538e57